Feature/login

cf2080 opened the pull request:
# Description
Things Changed/Added:
1. Added Httpost login method and Login Request Class in User Controller
2. Added VerifyPassword method in Password Service class in order to verify whether the password inputted is correct
3. Added IncorrectPasswordCount attribute to User model class in order to keep track of password attempts so a user can be disabled
4. Added cors to webapiconfig class
5. Added a simple login page to the front just for testing purposes
## Type of change
- [ ] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
- [ ] The code hasn't been unit tested but will add some soon

# Notes

Put any notes here

TriforceSoftwareDevelopment committed:
Created UI to test out login. VerfiedPassword method created in Password service class. Moved Login [httpost] method to User Controller. Also enabled cors.

cf2080 committed:
Added Incorrect Password Account to User Class. Added code in the controller to check the number of times the user incorrectly input their password.

julianpoy commented on Backend/DataAccessLayer/Migrations/201903062125096_initial-create.cs:
```diff
@@ -3,10 +3,24 @@ namespace DataAccessLayer.Migrations
     using System;
```
Hmm, why was this old migration moved or changed? Migrations tend to be treated as historical artifacts

julianpoy replied:
If you're trying to add a new field, make a new migration to modify the users table.

cf2080 replied:
accidentally deleted migration files when trying to connect to the database. @julianpoy

julianpoy replied:
Yeah, these need to be reversed before this can be resolved

cf2080 replied:
files are now reversed

julianpoy commented on Backend/DataAccessLayer/Migrations/Configuration.cs:
```diff
@@ -9,7 +9,7 @@ internal sealed class Configuration : DbMigrationsConfiguration<DataAccessLayer.
         public Configuration()
         {
-            AutomaticMigrationsEnabled = false;
+            AutomaticMigrationsEnabled = true;
```
Revert this unless we're making a policy change to have automatic migrations.

julianpoy commented on Backend/DataAccessLayer/Migrations/201903062125096_initial-create.resx:
```diff
@@ -118,7 +118,7 @@
   </resheader>
   <data name="Target" xml:space="preserve">
-    <value>H4sIAAAAAAAEAO1czXLbNhC+d6bvwOGp7TimLR/aeqRkHNlJPY3jNHIyvWVgEpIxJUGWAB15On2yHvpIfYUu+A8CpPgnWW46vkj4+bAAPuwusCv/89ff0xdrzzXucciIT2fm8eGRaWBq+w6hq5kZ8eWzH8wXz7/+anrheGvjY9buRLSDnpTNzDvOg1PLYvYd9hA79Igd+sxf8kPb9yzk+Nbk6OhH6/jYwgBhApZhTN9HlBMPx1/g69ynNg54hNwr38EuS8uhZhGjGm+Rh1mAbDwzzxFHZ7aNGXuDHnB4KL7fIoZN48wlCORZYHdpGohSnyMO0p5+YHjBQ5+uFgEUIPfmIcDQbolc0SuexWnRvO2EjiZiQlbRMYOyI8Z9ryPg8Um6 ...
+    <value>H4sIAAAAAAAEAOVazW7jNhC+F+g7CDq1RTZK7EPbwN5F4iRbo5t4G2UXvS1oaWwTpSgtSWVtFH2yHvpIfYUO9Wf9OpL/sEHhi01yvhkOZ4bDGf/79z+DN0uPGU8gJPX50Dw/PTMN4I7vUj4fmqGavfrJfPP6228GN663ND6m6/p6HVJyOTQXSgUXliWdBXhEnnrUEb70Z+rU8T2LuL7VOzv72To/twAhTMQyjMFDyBX1IPqBP0c+dyBQIWF3vgtMJuM4Y0eoxj3xQAbEgaF5TRS5dByQ8h1ZgTjVv6dEgmlcMkpQHhvYzDQI574iCqW9+CDBVsLnczvAAcIeVwHguhlhmiraxcV6edsNnfX0hqw1YQrlhFL5XkfA836iIatMvpWe ...
```
Revert this

julianpoy commented on Backend/DataAccessLayer/Models/User.cs:
```diff
@@ -39,6 +39,10 @@ public User()
         public string SecurityQ3Answer { get; set; }
+
+        [Required]
+        public int IncorrectPasswordCount { get; set; }
```
If this is non-nullable, make sure the user constructor for this model sets a default value (0)

julianpoy commented on Backend/Dev/App.config:
```diff
@@ -1,11 +1,11 @@
+  <!-- For more information on Entity Framework configuration, visit http://go.microsoft.com/fwlink/?LinkID=237468 --></configSections>
   <startup>
-    <supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.6.2" />
+    <supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.6.1" />
```
No downgrade here. Please revert this file.

julianpoy commented on Backend/Dev/Dev.csproj:
```diff
@@ -8,10 +8,11 @@
     <RootNamespace>Program</RootNamespace>
     <AssemblyName>Dev</AssemblyName>
-    <TargetFrameworkVersion>v4.6.2</TargetFrameworkVersion>
+    <TargetFrameworkVersion>v4.6.1</TargetFrameworkVersion>
```
Same as above

julianpoy commented on Backend/Dev/packages.config:
```diff
@@ -1,4 +1,4 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
-  <package id="EntityFramework" version="6.2.0" targetFramework="net462" />
+  <package id="EntityFramework" version="6.2.0" targetFramework="net461" />
```
Same as above. I'm going to stop commenting on this issue, please check all other instances of this change.

julianpoy commented on Backend/ServiceLayer/Services/PasswordService.cs:
```diff
@@ -59,5 +59,18 @@ public string[] QueryPwnedApi(string prefix,string url)
+        public bool VerifyPassword(string password, string hash, byte[] salt)
+        {
+            var rfc2898DeriveBytes = new Rfc2898DeriveBytes(password, salt);
+            return Convert.ToBase64String(rfc2898DeriveBytes.GetBytes(256)) == hash;
```
Hmm - I believe this functionality exists somewhere.

cf2080 replied:
yes, functionality is in user repository class. I will use that one instead.

julianpoy commented on Backend/UnitTesting/ITestingUtils.cs:
```diff
@@ -12,9 +12,9 @@ interface ITestingUtils
+        //Claim createClaim(User user, Service service);
         bool isEqual(string[] arr1, string[] arr2);
-        Client createClientObject();
+        //Client createClientObject();
```
What is the reason for these changes?

alfredovargasjr replied:
These are creating to project build errors, these classes do not existing in the solution

julianpoy commented on Backend/UnitTesting/UserServiceUT.cs:
```diff
@@ -381,7 +381,7 @@ public void Is_User_Manager_Over_Independent()
             using (_db = tu.CreateDataBaseContext())
             {
-                Assert.IsFalse(us.IsManagerOver(_db, unassociatedUser, subject));
+                //Assert.IsFalse(us.IsManagerOver(_db, unassociatedUser, subject));
```
Please remove these tests completely before PR. These shouldn't exist in the system anymore (applies to tests you commented below as well).

jbcayton replied:
Do you want the class deleted or just those specific tests?

alfredovargasjr replied:
This has been fixed on the bug branch

julianpoy commented on Backend/WebAPI/App_Start/WebApiConfig.cs:
```diff
@@ -10,6 +12,17 @@ public static class WebApiConfig
+
+            // Set JSON formatter as default one and remove XmlFormatter
+            var jsonFormatter = config.Formatters.JsonFormatter;
+            jsonFormatter.SerializerSettings.ContractResolver = new CamelCasePropertyNamesContractResolver();
```
:+1:

julianpoy commented on Frontend/src/router.js:
```diff
@@ -25,6 +27,16 @@ let router = new VueRouter({
+    {
+      path: '/portal',
+      name: 'portal',
+      component: Portal
```
Let's call this dashboard instead of portal.

julianpoy commented on Frontend/src/views/Login.vue:
```diff
@@ -0,0 +1,34 @@
+        },
+        methods: {
+            login() {
+               axios.post('http://localhost:50803/api/users/login').then(input => {this.input = input.data; console.log("sent!");})
```
This could use some better result/error handling.

julianpoy commented:
There are a giant number of bootstrap and other frontend files added to the backend project. Please remove those.

TheXulis committed:
Removed all calls to nonexistent classes from the Unit Tests

alfredovargasjr committed:
added fixes to webapi project and cleaned up some things

jbcayton committed:
Pull requests corrections

Automatic migrations disabled.
IncorrectPasswordCount default value set to 0.
Framework reverted back to 4.6.2
Testing file changes.
Portal page renamed to dashboard.

alfredovargasjr commented on Backend/DataAccessLayer/Database/dbContext.cs:
```diff
@@ -14,12 +14,10 @@ public class DatabaseContext : DbContext
         public DatabaseContext()
         {
-            this.Database.Connection.ConnectionString = "Data Source=localhost;Initial Catalog=SSO;Integrated Security=True";
+            Database.Connection.ConnectionString = "Data Source=Localhost;Initial Catalog=SSO;Integrated Security=True";
```
Why was this change made?

cf2080 replied:
just for testing purposes when I was trying to connect to the database. I'll revert back to the original format

alfredovargasjr commented on Backend/WebAPI/Areas/HelpPage/Views/Web.config:
```diff
@@ -9,7 +9,7 @@
   </configSections>
 
   <system.web.webPages.razor>
-    <host factoryType="System.Web.Mvc.MvcWebRazorHostFactory, System.Web.Mvc, Version=5.2.4.0, Culture=neutral, PublicKeyToken=31BF3856AD364E35" />
```
This WebAPI project will be replaced by a clean Web Api project that gets rid of unnecessary MVC packages. Refer to bug branch on the SSO repo

alfredovargasjr commented on Backend/WebAPI/Controllers/UsersController.cs:
```diff
@@ -65,6 +71,45 @@ public string Register([FromBody] RegisterRequest request)
+            IUserService _userService = new UserService();
+            IPasswordService _passwordService = new PasswordService();
+
+            using (var _db = new DatabaseContext())
```
This logic can be moved to the manager layer. If we feel like you can do it by the end of the milestone, go for it. If not, do it after so we can get functionality out of the way.

cf2080 replied:
It's done. Can you check it? @alfredovargasjr

alfredovargasjr commented on Backend/WebAPI/Controllers/UsersController.cs:
```diff
@@ -65,6 +71,45 @@ public string Register([FromBody] RegisterRequest request)
+                bool isPasswordMatched = _passwordService.VerifyPassword(request.password, user.PasswordHash, user.PasswordSalt);
+
+                //succesful login
+                if (isPasswordMatched == true)
```
This needs to return a token via a created session

cf2080 replied:
if statement now returns token through a created session, I believe.

alfredovargasjr commented on Backend/DataAccessLayer/Migrations/201903062125096_initial-create.cs:
```diff
@@ -3,10 +3,24 @@ namespace DataAccessLayer.Migrations
+                c => new
+                    {
+                        Id = c.Guid(nullable: false),
+                        Token = c.String(nullable: false),
```
Token should be a GUID

alfredovargasjr replied:
Another team has implemented a Token Service

cf2080 committed:
pull request corrections

cf2080 committed:
Merge branch 'bug/fixing-scaffold-error' into feature/login

cf2080 committed:
pull request corrections part 2

cf2080 committed:
Pull Request Corrections

- token is returned if login succesful in controller
- Created token Interface and service (took from reset password branch)

TheXulis committed:
There were just some outdated and missing NuGet packages

jbcayton committed:
Pull Request Correction

Moved logic for login to manager layer
Created model for LoginRequest

cf2080 committed:
Error Handling for axios in the front end. Created DTO folder in DataAccessLayer for LoginRequest class

cf2080 committed:
Refactoring

cf2080 committed:
Update DataAccessLayer.csproj

jbcayton committed:
Pull request correction

Updated password validation on LoginCheckPassword

cf2080 committed:
Update IPasswordService.cs

cf2080 committed:
Merge branch 'bug/fixing-scaffold-error' into feature/login

cf2080 committed:
_db.savechange() added in a few functions

cf2080 committed:
Merge branch 'master' into feature/login

alfredovargasjr commented on Backend/KFC-WebAPI/Controllers/UsersController.cs:
```diff
@@ -11,6 +11,7 @@
 using System.Net.Http;
 using System.Web.Http;
 using System.Web.Http.Cors;
+using DataAccessLayer.Repositories;
```
Why is this being called in the controller class?

cf2080 replied:
I was using it when I had the majority of the logic in the class. I'm going to delete it.

alfredovargasjr commented on Backend/DataAccessLayer/Models/LoginRequest.cs:
```diff
@@ -0,0 +1,14 @@
+using System.Threading.Tasks;
+
+namespace DataAccessLayer.Models
+{
```
Any DTO's that are used for controllers should be in the Models folder inside the web-api project. KFC-WebApi/Models

cf2080 replied:
Ok. I have a folder named DTOs in the DataAccessLayer, I'll delete it and move my DTO to the model's folder.

alfredovargasjr commented on Backend/KFC-WebAPI/Controllers/UsersController.cs:
```diff
@@ -63,6 +68,40 @@ public string Register([FromBody] RegisterRequest request)
+                {
+                    if(loginM.LoginCheckPassword(request.password))
+                    {
+                        return Ok(loginM.LoginAuthorized());
```
you should just return the token, not the entire session's object

cf2080 replied:
the object does return a token

alfredovargasjr replied:
Got it. The var response was never used in the manager class, i was confused lol.

alfredovargasjr commented on Backend/KFC-WebAPI/Controllers/UsersController.cs:
```diff
@@ -63,6 +68,40 @@ public string Register([FromBody] RegisterRequest request)
+            LoginManager loginM = new LoginManager();
+            if (loginM.LoginCheckUserExists(request.email) == false)
+            {
+                return NotFound();
```
Return some information so the error can be known. "Invalid username or password"

cf2080 replied:
I changed it with **return content(httpstatuscode.notfound,"message");**

alfredovargasjr commented on Backend/KFC-WebAPI/Scripts/index.d.ts:
```diff
@@ -0,0 +1,159 @@
+ * @fileoverview This file only declares the public portions of the API.
+ * It should not define internal pieces such as utils or modifier details.
+ *
+ * Original definitions by: edcarroll <https://github.com/edcarroll>, ggray <https://github.com/giladgray>, rhysd <https://rhysd.github.io>, joscha <https://github.com/joscha>, seckardt <https://github.com/seckardt>, marcfallows <https://github.com/marcfallows>
```
This is not supposed to be in here

alfredovargasjr replied:
KFC-WebAPI/Scripts

cf2080 replied:
I merged the branch 2 days ago. I don't have anymore

cf2080 replied:
with the bug branch

alfredovargasjr commented on Backend/KFC-WebAPI/Scripts/README.md:
```diff
@@ -0,0 +1,219 @@
+<!-- IGNORE THE HTML BLOCK BELOW, THE INTERESTING PART IS AFTER IT -->
```
This has been removed by the bug branch. Accept the changes from the bug branch.

cf2080 replied:
I merged the branch 2 days ago. I don't have anymore

alfredovargasjr replied:
on that commit, you accepted some stuff but still seems like you kept old files.

jbcayton committed:
Pull request changes

jbcayton committed:
Pull request corrections and unit test

Changed return type of login to token
Set up unit test for LoginManager

cf2080 committed:
deleted scripts folder

cf2080 committed:
Update LoginRequest.cs

cf2080 committed:
Update UsersController.cs

jbcayton committed:
Unit testing

jbcayton committed:
Merge branch 'feature/login' of https://github.com/CECS-491A/SSO into feature/login

cf2080 committed:
Unit test example

cf2080 committed:
Merge branch 'feature/login' of https://github.com/CECS-491A/SSO into feature/login

cf2080 committed:
Update LoginManagerUT.cs

alfredovargasjr commented on Backend/KFC-WebAPI/Content/bootstrap-grid.css:
```diff
@@ -0,0 +1,3719 @@
+/*!
```
This also has to be removed, this is irrelevant for the web api project.

alfredovargasjr commented on Backend/KFC-WebAPI/Content/bootstrap-reboot.css:
```diff
@@ -0,0 +1,331 @@
+/*!
```
Delete KFC-WebApi/content

cf2080 committed:
Aded error handling to login method

cf2080 committed:
Change context name to original one

julianpoy commented on .vs/ProjectSettings.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "CurrentProjectSetting": null
```
This file should be removed

julianpoy replied:
All other .vs files should be removed below

julianpoy commented on Backend/DataAccessLayer/App.config:
```diff
@@ -1,21 +1,20 @@
   <connectionStrings>
-    <add name="ServerDB" providerName="System.Data.SqlClient"
-          connectionString="Data Source=localhost;Initial Catalog=SSO;Integrated Security=True" />
+    <add name="ServerDB" providerName="System.Data.SqlClient" connectionString="Data Source=localhost;Initial Catalog=SSO;Integrated Security=True"/>
```
This file should not be changed

julianpoy commented on Backend/DataAccessLayer/DataAccessLayer.csproj:
```diff
@@ -49,15 +50,12 @@
-      <DependentUpon>201812040642468_create-table-session.cs</DependentUpon>
+    <Compile Include="Migrations\201903062125096_initial-create.cs" />
+    <Compile Include="Migrations\201903062125096_initial-create.Designer.cs">
+      <DependentUpon>201903062125096_initial-create.cs</DependentUpon>
```
This migration should not be added, and the above migrations should not be removed

julianpoy commented on Backend/DataAccessLayer/DataAccessLayer.csproj:
```diff
@@ -75,13 +73,12 @@
-    <EmbeddedResource Include="Migrations\201812040642468_create-table-session.resx">
-      <DependentUpon>201812040642468_create-table-session.cs</DependentUpon>
+    <EmbeddedResource Include="Migrations\201903062125096_initial-create.resx">
+      <DependentUpon>201903062125096_initial-create.cs</DependentUpon>
```
Undo this

cf2080 replied:
done

julianpoy commented on Backend/DataAccessLayer/DataAccessLayer.csproj:
```diff
@@ -75,13 +73,12 @@
-  <ItemGroup />
+  <ItemGroup>
+    <Folder Include="DTOs\" />
+  </ItemGroup>
```
Why this change?

cf2080 replied:
I created a folder to put the LoginRequest class there but decided to move it into the model class

julianpoy commented on Backend/DataAccessLayer/Database/DatabaseContext.cs:
```diff
@@ -16,10 +16,8 @@ public DatabaseContext()
         public DbSet<User> Users { get; set; }
         public DbSet<Session> Sessions { get; set; }
-
-
```
Undo all changes to this file

julianpoy replied:
All designer files below this one that have been removed need to be re-added

cf2080 committed:
Update DataAccessLayer.csproj

julianpoy commented on Backend/DataAccessLayer/Migrations/201903062125096_initial-create.cs:
```diff
@@ -3,10 +3,24 @@ namespace DataAccessLayer.Migrations
+                    })
+                .PrimaryKey(t => t.Id)
+                .ForeignKey("dbo.Users", t => t.UserId, cascadeDelete: true)
+                .Index(t => t.UserId);
```
Undo this block of green. It should all be unchanged, including the class name of this file.

julianpoy commented on Backend/DataAccessLayer/Migrations/201903062125096_initial-create.cs:
```diff
@@ -26,17 +40,21 @@ public override void Up()
+            DropIndex("dbo.Sessions", new[] { "UserId" });
             DropTable("dbo.Users");
-
+            DropTable("dbo.Sessions");
```
Undo this change

julianpoy commented on Backend/DataAccessLayer/Migrations/201903062125096_initial-create.cs:
```diff
@@ -26,17 +40,21 @@ public override void Up()
                         SecurityQ2Answer = c.String(),
                         SecurityQ3 = c.String(),
                         SecurityQ3Answer = c.String(),
+                        IncorrectPasswordCount = c.Int(nullable: false),
```
This needs to be in it's own new migration. You can't modify old migrations.

cf2080 committed:
Revert completely to main branch.

cf2080 committed:
Added login implementation after reverting migration files

alfredovargasjr commented:
@cf2080 resolve conflicts

cf2080 committed:
Merge branch 'master' into feature/login

cf2080 committed:
testing

julianpoy commented on Frontend/src/views/Login.vue:
```diff
@@ -0,0 +1,60 @@
+               {
+                  'Access-Control-Allow-Origin': '*',
+                  'Access-Control-Allow-Credentials': true
+                })
```
These headers should not be necessary.

cf2080 committed:
Update Login.vue
